crates/biome_js_analyze/src/lint/nursery/no_magic_numbers.rs

```diff
@@ -0,0 +1,239 @@
+    /// let total = price * 1.23; // Magic number for tax rate
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const TAX_RATE = 1.23;
+    /// let total = price * TAX_RATE;
+    /// ```
+    pub NoMagicNumbers {
+        version: "next",
+        name: "noMagicNumbers",
+        language: "js",
+        sources: &[RuleSource::Eslint("no-magic-numbers")],
+        recommended: false,
+    }
+}
+
+#[derive(Debug)]
+pub struct NumberContext {
+    range: TextRange,
+    raw: String,
+    is_const_violation: bool,
+}
+
+impl Rule for NoMagicNumbers {
+    type Query = Ast<AnyJsExpression>;
+    type State = NumberContext;
+    type Signals = Option<Self::State>;
+    type Options = NoMagicNumbersConfig;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let config = ctx.options();
+
+        match node {
+            AnyJsExpression::AnyJsLiteralExpression(literal_expr) => {
+                if let Some(literal) = literal_expr.as_js_number_literal_expression() {
+                    let value_token = literal.value_token().ok()?;
+                    let value_text = value_token.text();
+                    let value = value_text.parse::<i64>().ok()?;
+                    if is_allowed_number(literal.syntax(), value, config) {
+                        return None;
+                    }
+
+                    let is_const_violation = check_const_violation(literal.syntax(), config);
+
+                    Some(NumberContext {
+                        range: literal.range(),
+                        raw: value_token.text().to_string(),
+                        is_const_violation,
+                    })
+                } else {
+                    return None;
+                }
+            }
+            AnyJsExpression::JsUnaryExpression(unary) => {
+                if let Ok(operator) = unary.operator() {
+                    if operator == JsUnaryOperator::Minus {
+                        if let Ok(argument) = unary.argument() {
+                            if let Some(literal_expr) = argument.as_any_js_literal_expression() {
+                                if let Ok(value_token) = literal_expr.value_token() {
+                                    let value_text = value_token.text();
+                                    let value = value_text.parse::<i64>().ok()?;
+                                    let neg_value = -value;
+                                    if is_allowed_number(unary.syntax(), neg_value, config) {
+                                        return None;
+                                    }
+                                    let is_const_violation =
+                                        check_const_violation(unary.syntax(), config);
+                                    return Some(NumberContext {
+                                        range: unary.range(),
+                                        raw: format!("-{}", value_token.text()),
```

Comment:
Technically, this isn't the raw value because you're adding a `-`. Plus, you already have `value_text`, use that one